In [1]:
import pandas as pd 
import numpy as np

In [2]:
path = '../ml-latest/'
filename = 'ratings.csv'

In [3]:
data = pd.read_csv(path+filename)

Let us try and see now how many unique users are there

In [4]:
user_column_name = 'userId'
item_column_name = 'movieId'
rating_column_name = 'rating'

In [5]:
len(set(data[user_column_name]))

283228

In [6]:
len(set(data[item_column_name]))

53889

We see that we have $283228$ users and $53889$ movies. Let us now try and create the subset based on count and frequency of users watching a movie

In [29]:
users_data = data.groupby([user_column_name]).size().reset_index(name='number_of_movies_watched')

We now see the percentage of users who have seen at least a few movies

In [30]:
count_of_movies_watches = users_data.groupby(['number_of_movies_watched']).size().reset_index(name='Sum')

In [31]:
count_of_movies_watches['Percent'] = 100 * count_of_movies_watches['Sum']/np.sum(count_of_movies_watches['Sum'])

In [32]:
count_of_movies_watches['CumSum'] = 100 - count_of_movies_watches['Percent'].cumsum()

In [35]:
count_of_movies_watches.head(30)

number_of_movies_watched    Sum   Percent     CumSum
0                          1   5620  1.984267  98.015733
1                          2   3951  1.394989  96.620744
2                          3   3534  1.247758  95.372986
3                          4   3143  1.109707  94.263279
4                          5   7614  2.688294  91.574986
5                          6   5615  1.982502  89.592484
6                          7   4220  1.489966  88.102518
7                          8   3183  1.123830  86.978689
8                          9   2690  0.949765  86.028924
9                         10   7000  2.471507  83.557417
10                        11   4597  1.623074  81.934343
11                        12   3812  1.345912  80.588431
12                        13   3305  1.166904  79.421526
13                        14   3009  1.062395  78.359131
14                        15  19545  6.900801  71.458330
15                        16  10266  3.624642  67.833689
16                        17   7125  2.515641  65.318048
17                        18   5708  2.015337  63.302710
18                        19   4686  1.654497  61.648213
19                        20   5052  1.783722  59.864491
20                        21   3705  1.308133  58.556357
21                        22   3368  1.189148  57.367209
22                        23   3147  1.111119  56.256090
23                        24   2894  1.021792  55.234299
24                        25   2914  1.028853  54.205446
25                        26   2642  0.932817  53.272628
26                        27   2488  0.878444  52.394184
27                        28   2447  0.863968  51.530216
28                        29   2402  0.848080  50.682136
29                        30   2357  0.832192  49.849944

From the above analysis, we see that $91%$ of the users have watched at least $5$ movies and $83%$ of the users have seen at least $10$ movies

$50%$ of the users have seen at least $29$ movies. If we want to remove sparcity, we would need to identify the threshold that we require to use our model. I believe the ideal number would be between $5$ and $10$ 

Similarly, we need to see how to stratify the items

In [37]:
items_data = data.groupby([item_column_name]).size().reset_index(name='number_of_users_seen')
count_of_items_watches = items_data.groupby(['number_of_users_seen']).size().reset_index(name='Sum')
count_of_items_watches['Percent'] = 100 * count_of_items_watches['Sum']/np.sum(count_of_items_watches['Sum'])
count_of_items_watches['CumSum'] = 100 - count_of_items_watches['Percent'].cumsum()

In [39]:
count_of_items_watches.head(10)

number_of_users_seen    Sum    Percent     CumSum
0                     1  10155  18.844291  81.155709
1                     2   6324  11.735234  69.420475
2                     3   3919   7.272356  62.148119
3                     4   2667   4.949062  57.199057
4                     5   2033   3.772570  53.426488
5                     6   1674   3.106385  50.320102
6                     7   1321   2.451335  47.868767
7                     8   1099   2.039377  45.829390
8                     9    981   1.820409  44.008981
9                    10    860   1.595873  42.413108

We see that $50%$ of the movies have been consumed by at least $6$ users. There are also $18%$ of the movies that have only been seen by $1$ user. These items should not be used for training since we would not be able to learn from them. We should look at movies which have been seen by at least $4$ users which covers $57%$ of the movies 

Creating a subset based on the baove combinations would need to be discussed. 

ImDB ID is present in the file links.csv. The leading zeroes in that file are removed when opening in Excel. Need to make sure that we read the column as a string in python. 

Till now the if the IMDB ID is 0114709, then looking at the link https://www.imdb.com/title/tt + 0114709/ is opening up the ImDB page. Maybe we can use that to extract information from ImDB. 